### PR TITLE
Improve transform error handling

### PR DESCRIPTION
**Pull request by jan-elastic**
https://github.com/elastic/ml-team/issues/1057

### CODE REVIEW

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java
Comment (jan-elastic) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java:508]: Adding just the exception to the original code:
```
logger.error("[{}] transform has failed; experienced: [{}].", transform.getId(), reason, exception);
```
triggers the `ESLoggerUsageChecker`, hence this more verbose approach.

I think this is a shortcoming of the checker, see also:
https://www.slf4j.org/faq.html#paramException

Comment (droberts195) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java:508]: If I remember correctly that's not a shortcoming of the checker. The exception doesn't get logged with the simpler syntax because it ends up calling [this method](https://github.com/apache/logging-log4j2/blob/24f4fa990278a2346e8345d8fd4c09246451bd34/log4j-api/src/main/java/org/apache/logging/log4j/spi/AbstractLogger.java#L664) so the exception ends up as an unused message parameter.

The other way people work around this is:

```
logger.error(Strings.format("[{}] transform has failed; experienced: [{}].", transform.getId(), reason), exception);
```

File: docs/changelog/102240.yaml
Comment (droberts195) [inline docs/changelog/102240.yaml:2]:
```suggestion
summary: Exclude stack traces from transform audit messages and health
```

**Review by droberts195 — APPROVED** [review]
LGTM if you could just make the release note a bit more specific before merging

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java:60]: That's a nice finding :)
I've always struggled with message formatting when I passed the exception to `debug()`

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java:76]: Will the cause always be non-null?

**Review by przemekwitek — APPROVED** [review]
LGTM

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java
Comment (droberts195) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java:76]: I was just about to say that it cannot be `null` because that's enforced here: https://github.com/elastic/elasticsearch/blob/12ad399c488f0cc60e19b5e1b29c6d569cb4351a/server/src/main/java/org/elasticsearch/action/ShardOperationFailedException.java#L39

However, then I noticed that the exception we're treating specially here is `SearchPhaseExecutionException`. Should it be `ShardOperationFailedException` instead? That would be consistent with what was originally changed in #102200. But there are so many layers of wrapping and nesting that maybe by the time it gets here it's wrapped in a `SearchPhaseExecutionException`.

Comment (droberts195) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java:76]: There are 18 places in #45878 where `ExceptionsHelper.stackTrace()` is used. All of the ones that might affect transforms relate to searching shards.

If it's unclear whether to use `SearchPhaseExecutionException` or `ShardOperationFailedException` here then the safest thing might be to handle both, but account for the possibility that `e.getCause()` could be `null`.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java
Comment (jan-elastic) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java:508]: The following code:
```
        Exception e = new ElasticsearchException("EXCEPTION");
        Logger logger = LogManager.getLogger(Elasticsearch.class);
        logger.error("zero params + exception", e);
        logger.error("one param [{}] + exception", 42, e);
        logger.error("two params [{}] [{}]", 42, e);
        logger.error("two params + exception [{}] [{}]", 42, e, e);
```

results in:

```
» [2023-11-16T10:29:26,024][ERROR][o.e.b.Elasticsearch      ] [runTask-0] zero params + exception org.elasticsearch.ElasticsearchException: EXCEPTION
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase1(Elasticsearch.java:143)
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:69)
»  
» [2023-11-16T10:29:26,043][ERROR][o.e.b.Elasticsearch      ] [runTask-0] one param [42] + exception org.elasticsearch.ElasticsearchException: EXCEPTION
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase1(Elasticsearch.java:143)
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:69)
»  
» [2023-11-16T10:29:26,044][ERROR][o.e.b.Elasticsearch      ] [runTask-0] two params [42] [org.elasticsearch.ElasticsearchException: EXCEPTION]
» [2023-11-16T10:29:26,044][ERROR][o.e.b.Elasticsearch      ] [runTask-0] two params + exception [42] [org.elasticsearch.ElasticsearchException: EXCEPTION] org.elasticsearch.ElasticsearchException: EXCEPTION
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase1(Elasticsearch.java:143)
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:69)
```

As you can see: an additional trailing param of the type exception is always logged.

Comment (jan-elastic) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java:508]: This is the behavior since SLF4J 1.6.0 (dating back till 2010). Before that, the exception would be ignored in these cases, see the link above.

Comment (jan-elastic) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java:508]: Regarding:
```
logger.error(Strings.format("[{}] transform has failed; experienced: [{}].", transform.getId(), reason), exception);
```

That works indeed, but has some disadvantages:
- Some logging implementations collect the unexpanded pattern-string as a queryable field.
- It also prevents spending time rendering a message that no logger is configured to output.

(just quoting https://google.github.io/styleguide/pyguide.html#3101-logging)

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java
Comment (jan-elastic) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java:76]: If you look at the [Github issue](https://github.com/elastic/ml-team/issues/1057), the exception message starts with
```
Failed to execute phase [
```
which indicates a `SearchPhaseExecutionException`.

The message contains submessages of the shards that failed (each between curly brackets). These nested `ShardSearchFailure`s contain the stack traces.

Comment (jan-elastic) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java:76]: Good catch about the non-null. 

Looking at the code I'd say it's non-null, because the `SearchPhaseExecutionException`'s implementation of `getCause` processes the nested `ShardOperationFailedException`s, which have non-null causes. But the code is non-trivial, I'm not 100% sure.

Comment (jan-elastic) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformFailureHandler.java]: I've added a null check, just in case

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java
Comment (droberts195) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java:508]: That's interesting. It does look like `ESLoggerUsageChecker` is out-of-date. I know back in about 2017 we were having problems with exceptions getting lost from log statements. Elasticsearch was using log4j version 1 then I think though, so that was probably why. I guess the checker didn't get upgraded when log4j was made more intuitive.

Comment (jan-elastic) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java:508]: Any advice on what to do next?

- file bug report on ESLoggerUsageChecker (and where?)
- do nothing
- fix it myself

**Pull request merged**
